### PR TITLE
Async State Hashing

### PR DESCRIPTION
**Pull request by d4hines**
## Depends

- [x] #413
- [x] #412
- [x] #501 

## Problem

Adds full async hashing per #147 . 

## Solution

In #412, we started storing a list of snapshots. In this PR, we instead store a list of mutable references to `option(snapshot)`'s. 
We hash the protocol's on different threads using `Domainslib.Task.async`,and then updates the references when done.

**This assumes Atomic.t works out of the box with multicore! Is this true?**

Reviewers probably want to check that I'm setting up `Domainslib.Task.setup_pool` correctly. I discovered that unless I call it after `Lwt_main.run`, `Lwt_main.run` fails with:
```
Unix.fork may not be called while other domains were created
```

There is some new complexity in `deku_node.ml`. Specifically, because the `snapshot.current_snapshot` may be `None`, we keep a ref of the last snapshot available, so we can always return a snapshot to those who ask. I believe it's important to always have the latest snapshot available to other nodes to minimize synchronization time. But if you can think of a cleaner way to do it than what I've done, please let me know.

This is still missing the recovery mechanism from #250 - I'll rebase that PR to this one.

## Related
- Supersedes #243

### CODE REVIEW

File: bin/deku_node.re [hunk 1]
```diff
@@ -132,19 +134,34 @@ let handle_block_level =
     Ok({level: Flows.find_block_level(Server.get_state())})
   });
 
+let empty_hash = Snapshots.{hash: BLAKE2B.hash(""), data: ""};
```
Comment (EduardoRFS) [inline bin/deku_node.re:137]:
```suggestion
// TODO: this clearly should not be here
let empty_hash = Snapshots.{hash: BLAKE2B.hash(""), data: ""};
```

File: bin/deku_node.re [hunk 2]
```diff
@@ -132,19 +134,34 @@ let handle_block_level =
...
-      });
+      let.ok Snapshots.{hash: snapshot_hash, data: snapshot} =
+        switch (Snapshots.get_current_snapshot(snapshots)) {
+        | Some(snapshot) =>
+          latest_snapshot := (true, snapshot);
+          Ok(snapshot);
+        | None =>
+          let (initialized, latest_snapshot) = latest_snapshot^;
+          let.assert () = (`Node_not_yet_initialized, initialized);
+          Ok(latest_snapshot);
```
Comment (EduardoRFS) [inline bin/deku_node.re:153]: This is a problem because of

https://github.com/marigold-dev/deku/blob/68170bf7315bb88493b86e6d7314a6047c495468/node/state.re#L191

File: node/snapshots.rei [hunk 3]
```diff
@@ -1,24 +1,43 @@
...
-let update:
-  (~new_snapshot: (BLAKE2B.t, string), ~applied_block_height: int64, t) => t;
+let add_snapshot_ref: (~block_height: int64, t) => (snapshot_ref, t);
+
+let set_snapshot_ref: (snapshot_ref, snapshot) => unit;
+
+/** Combines add_snapshot_ref and set_snapshot_ref in a single operation.
+    Used when synchronously hashing a protocol state.
+*/
+let add_snapshot: (~block_height: int64, snapshot, t) => t;
```
Comment (EduardoRFS) [inline node/snapshots.rei:37]: I would prefer if this function was removed, it is effectively only an util as it doesn't wrap considerable logic

File: node/state.re [hunk 4]
```diff
@@ -243,7 +200,36 @@ let load_snapshot =
...
+  List.fold_left_ok(
+    (prev_state, block) => {
+      let.ok state = apply_block(prev_state, block);
+      if (BLAKE2B.equal(
+            prev_state.protocol.state_root_hash,
+            state.protocol.state_root_hash,
+          )) {
+        Ok(state);
+      } else {
+        let (hash, data) = Protocol.hash(prev_state.protocol);
```
Comment (EduardoRFS) [inline node/state.re:222]: This should definitely be in parallel. Otherwise in a tight situation the node will not be able to load the snapshot fast enough.

File: node/flows.re [hunk 5]
```diff
@@ -246,19 +248,32 @@ let try_to_commit_state_hash = (~prev_validators, state, block, signatures) => {
...
 };
 
+let hash_new_state_root = (state, protocol, update_state) => {
+  let task_pool = get_task_pool^();
+  let (snapshot_ref, snapshots) =
+    Snapshots.add_snapshot_ref(
+      ~block_height=protocol.block_height,
+      state.Node.snapshots,
+    );
+  let _ =
```
Comment (EduardoRFS) [inline node/flows.re:258]: with async stuff is very easy to miss a promise or a domain by accident, at least do `(_task: type)`

Comment (d4hines) [reply node/flows.re]: Done

File: node/state.re (hunk 4 above)
Comment (d4hines) [reply node/state.re]: I added a `TODO:`

File: node/snapshots.rei (hunk 3 above)
Comment (d4hines) [reply node/snapshots.rei]: Done

**Comment by d4hines** [comment]
Rebased with main and tested this branch today.

**Comment by d4hines** [comment]
A manual test I did: inject a `Unix.sleep 5` in the function executed on the other core with `Task.async` in the `hash_new_state_root` function of `flows.ml`. I then observed that this sleep did not cause any visible slow down in block production.

File: src/node/flows.ml [hunk 6]
```diff
@@ -206,14 +209,21 @@ let try_to_commit_state_hash ~prev_validators state block signatures =
...
+  update_state { state with snapshots }
+
 let rec try_to_apply_block state update_state block =
   let%assert () =
     ( `Block_not_signed_enough_to_apply,
       Block_pool.is_signed ~hash:block.Block.hash state.Node.block_pool ) in
-  let%assert () =
-    ( `Invalid_state_root_hash,
-      block_matches_current_state_root_hash state block
-      || block_matches_next_state_root_hash state block ) in
```
Comment (EduardoRFS) [inline src/node/flows.ml:216]: Sorry, why is this removed again? Shouldn't we always have the next state root hash? And if it's late it should hash it synchronously IIRC.

File: src/bin/deku_node.ml [hunk 7]
```diff
@@ -58,19 +58,30 @@ let handle_block_level =
...
-            Signatures.to_list snapshots.last_block_signatures;
-        })
+        Networking.Protocol_snapshot.
+          {
+            snapshot;
+            additional_blocks = snapshots.additional_blocks;
+            last_block = snapshots.last_block;
+            last_block_signatures =
+              Signatures.to_list snapshots.last_block_signatures;
+          })
```
Comment (EduardoRFS) [inline src/bin/deku_node.ml:84]:
```suggestion
          {
            snapshot;
            additional_blocks = snapshots.additional_blocks;
            last_block = snapshots.last_block;
            last_block_signatures =
              Signatures.to_list snapshots.last_block_signatures;
          })
```

File: src/bin/deku_node.ml [hunk 8]
```diff
@@ -58,19 +58,30 @@ let handle_block_level =
     (module Networking.Block_level)
     (fun _update_state _request ->
       Ok { level = Flows.find_block_level (Server.get_state ()) })
+let latest_snapshot = ref None
```
Comment (EduardoRFS) [inline src/bin/deku_node.ml:61]: Why isn't this on node state?

**Review by EduardoRFS — CHANGES_REQUESTED** [review]

File: src/node/flows.ml (hunk 6 above)
Comment (d4hines) [reply src/node/flows.ml:216]: If it's late, it does not start hashing asynchronously - this PR is fully async. 

The way it currently works is that if you miss a state root hash, we apply it anyway (because this check is removed), but you won't sign blocks. In #250, we'll add a way to sign those blocks if you later finish the state root hash. 

If we add a synchronous hashing mechanism, then there's no need to remove this check, and also no need for #250, but it may cause nodes to hash the same state on multiple cores. I think we could implement something more complex like "if you don't have the hash you need but it's being hashed on another core, wait until it's finished" - I'd have to think about it though.

File: src/node/server.ml [hunk 9]
```diff
@@ -29,6 +29,18 @@ let rec reset_timeout server =
...
      reset_timeout server;
      Lwt.return_unit)
+
+let task_pool = ref None
+
+let get_task_pool () =
+  match !task_pool with
+  | Some pool -> pool
+  | None ->
+    let pool = Domainslib.Task.setup_pool ~num_additional_domains:4 () in
```
[the remark this answers is not in the document]
Comment (aguillon) [reply src/node/server.ml:40]: @EduardoRFS should we explicitly require (and check for) a minimal number of cores in the future?

File: src/node/server.ml [hunk 10]
```diff
@@ -29,6 +29,19 @@ let rec reset_timeout server =
      | Error `Not_current_block_producer -> ());
      reset_timeout server;
      Lwt.return_unit)
+
+let task_pool = ref None
```
Comment (aguillon) [inline src/node/server.ml:33]: Not a fan, but I understand the why. I think the different parts of Deku's initialization should appear all together somewhere in the future (probably bin/node_state) to be more explicit. What do you think? @EduardoRFS @d4hines

**Review by aguillon — APPROVED** [review]
I have nothing against this PR. We'll discuss merging later via Slack.

File: src/node/flows.ml (hunk 6 above)
Comment (EduardoRFS) [reply src/node/flows.ml:216]: The check can be removed but we should add a comment here, this should make you go out of sync so that you don't sign blocks and then if you finish hashing it late and it matches you can go back to being in sync(as your hypothesis was correct).

File: src/node/server.ml (hunk 10 above)
Comment (EduardoRFS) [reply src/node/server.ml:33]: Agreed with Daniel. Almost all global state is initialized at server.ml, the remaining are state that are not relevant to us, such as the RNG.

File: src/bin/deku_node.ml (hunk 7 above)
Comment (EduardoRFS) [reply src/bin/deku_node.ml:75]: Same idea as above

File: src/node/server.ml (hunk 10 above)
Comment (aguillon) [reply src/node/server.ml:33]: Agreed.

**Review by aguillon — APPROVED** [review]

File: src/node/snapshots.ml [hunk 11]
```diff
@@ -73,7 +73,19 @@ let start_new_epoch t =
       additional_blocks;
     }
   | [] -> failwith "You must add a snapshot before you can start a new epoch"
-let get_current_snapshot t = Atomic.get t.current_snapshot
+
+let latest_finished_snapshot = ref None
```
Comment (EduardoRFS) [inline src/node/snapshots.ml:77]:
```ocaml
(* GLOBAL STATE *)
```
Please add a comment here as it's not in server.ml

File: src/node/snapshots.ml [hunk 12]
```diff
@@ -34,16 +41,20 @@ let append_block ~pool (block, signatures) t =
...
+  ( atom,
+    {
+      next_snapshots = t.next_snapshots @ [(block_height, atom)];
+      current_snapshot = t.current_snapshot;
+      last_block = t.last_block;
+      last_block_signatures = t.last_block_signatures;
+      additional_blocks = t.additional_blocks;
+    } )
+let set_snapshot_ref ref_ snapshot =
+  Format.printf "\027[36m New protocol snapshot hash: %s\027[m\n%!"
```
Comment (EduardoRFS) [inline src/node/snapshots.ml:55]:
```ocaml
Format.eprintf
```

**Review by EduardoRFS — APPROVED** [review]
anime girl with a tambourine approves this PR

File: src/node/flows.ml (hunk 6 above)
Comment (d4hines) [reply src/node/flows.ml:216]: added a comment in latest commit
